File: src/services/videoStudio.ts
// AI Video Studio / Colab Integration Service Interface and Implementation
export interface VideoProject {
  id: string;
  userId: string;
  title: string;
  description?: string;
  storyboard: StoryboardScene[];
  status: VideoProjectStatus;
  jobId?: string;
  jobConfig: Record<string, any>;
  resultUrls: string[];
  progress: number;
  errorMessage?: string;
  createdAt: Date;
  updatedAt: Date;
}

export type VideoProjectStatus = 'draft' | 'processing' | 'completed' | 'failed';

export interface StoryboardScene {
  id: string;
  title: string;
  description: string;
  durationSeconds: number;
  prompt: string;
  style?: string;
  referenceImages: string[];
  cameraAngle?: string;
  transition?: string;
  metadata: Record<string, any>;
}

export interface CreateVideoProjectRequest {
  title: string;
  description?: string;
  storyboard: StoryboardScene[];
  jobConfig: Record<string, any>;
}

export interface UpdateVideoProjectRequest {
  title?: string;
  description?: string;
  storyboard?: StoryboardScene[];
  status?: VideoProjectStatus;
}

export interface GenerateVideoRequest {
  projectId: string;
  config?: Record<string, any>;
}

export interface VideoGenerationStatus {
  projectId: string;
  status: VideoProjectStatus;
  progress: number;
  currentStep?: string;
  estimatedRemainingMinutes?: number;
  jobId?: string;
  startedAt?: string;
  updatedAt?: string;
  error?: string;
}

export interface VideoTemplate {
  id: string;
  name: string;
  description: string;
  scenes: StoryboardScene[];
  totalDuration: number;
  style: string;
}

export interface VideoGenerationQueue {
  activeJobs: number;
  queuedJobs: number;
  completedToday: number;
  averageProcessingTimeMinutes: number;
  estimatedWaitTimeMinutes: number;
  activeJobs: Array<{
    projectId: string;
    title: string;
    progress: number;
    startedAt: string;
    estimatedCompletion: string;
  }>;
}

export interface ListVideoProjectsQuery {
  userId?: string;
  status?: VideoProjectStatus;
  limit?: number;
  offset?: number;
}

export interface VideoStudioService {
  // Project management
  getVideoProjects: (query?: ListVideoProjectsQuery) => Promise<VideoProject[]>;
  getVideoProject: (id: string) => Promise<VideoProject>;
  createVideoProject: (request: CreateVideoProjectRequest) => Promise<VideoProject>;
  updateVideoProject: (id: string, request: UpdateVideoProjectRequest) => Promise<VideoProject>;
  deleteVideoProject: (id: string) => Promise<void>;
  
  // Video generation
  generateVideo: (request: GenerateVideoRequest) => Promise<{
    success: boolean;
    projectId: string;
    jobId: string;
    estimatedDurationMinutes: number;
  }>;
  getVideoGenerationStatus: (projectId: string) => Promise<VideoGenerationStatus>;
  cancelVideoGeneration: (projectId: string) => Promise<{
    success: boolean;
    message: string;
  }>;
  
  // File uploads
  uploadReferenceImage: (projectId: string, files: File[]) => Promise<{
    success: boolean;
    uploadedFiles: string[];
    projectId: string;
  }>;
  
  // Templates
  getVideoTemplates: () => Promise<VideoTemplate[]>;
  createProjectFromTemplate: (
    templateId: string,
    request: { title: string; description?: string }
  ) => Promise<VideoProject>;
  
  // Queue management
  getVideoGenerationQueue: () => Promise<VideoGenerationQueue>;
}

class VideoStudioServiceImpl implements VideoStudioService {
  private baseUrl: string;

  constructor(baseUrl: string = 'http://localhost:8080') {
    this.baseUrl = baseUrl;
  }

  async getVideoProjects(query?: ListVideoProjectsQuery): Promise<VideoProject[]> {
    const params = new URLSearchParams();
    if (query?.userId) params.append('user_id', query.userId);
    if (query?.status) params.append('status', query.status);
    if (query?.limit) params.append('limit', query.limit.toString());
    if (query?.offset) params.append('offset', query.offset.toString());

    const response = await fetch(`${this.baseUrl}/api/video-studio/projects?${params}`);
    if (!response.ok) throw new Error(`Failed to fetch video projects: ${response.statusText}`);
    return response.json();
  }

  async getVideoProject(id: string): Promise<VideoProject> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/${id}`);
    if (!response.ok) throw new Error(`Failed to fetch video project: ${response.statusText}`);
    return response.json();
  }

  async createVideoProject(request: CreateVideoProjectRequest): Promise<VideoProject> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to create video project: ${response.statusText}`);
    return response.json();
  }

  async updateVideoProject(id: string, request: UpdateVideoProjectRequest): Promise<VideoProject> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/${id}`, {
      method: 'PUT',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to update video project: ${response.statusText}`);
    return response.json();
  }

  async deleteVideoProject(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/${id}`, {
      method: 'DELETE',
    });
    if (!response.ok) throw new Error(`Failed to delete video project: ${response.statusText}`);
  }

  async generateVideo(request: GenerateVideoRequest): Promise<{
    success: boolean;
    projectId: string;
    jobId: string;
    estimatedDurationMinutes: number;
  }> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/${request.projectId}/generate`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to generate video: ${response.statusText}`);
    return response.json();
  }

  async getVideoGenerationStatus(projectId: string): Promise<VideoGenerationStatus> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/${projectId}/status`);
    if (!response.ok) throw new Error(`Failed to get video generation status: ${response.statusText}`);
    return response.json();
  }

  async cancelVideoGeneration(projectId: string): Promise<{
    success: boolean;
    message: string;
  }> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/${projectId}/cancel`, {
      method: 'POST',
    });
    if (!response.ok) throw new Error(`Failed to cancel video generation: ${response.statusText}`);
    return response.json();
  }

  async uploadReferenceImage(projectId: string, files: File[]): Promise<{
    success: boolean;
    uploadedFiles: string[];
    projectId: string;
  }> {
    const formData = new FormData();
    files.forEach((file, index) => {
      formData.append(`reference_image_${index}`, file);
    });

    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/${projectId}/upload`, {
      method: 'POST',
      body: formData,
    });
    if (!response.ok) throw new Error(`Failed to upload reference images: ${response.statusText}`);
    return response.json();
  }

  async getVideoTemplates(): Promise<VideoTemplate[]> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/templates`);
    if (!response.ok) throw new Error(`Failed to fetch video templates: ${response.statusText}`);
    return response.json();
  }

  async createProjectFromTemplate(
    templateId: string,
    request: { title: string; description?: string }
  ): Promise<VideoProject> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/projects/templates/${templateId}/from-template`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(request),
    });
    if (!response.ok) throw new Error(`Failed to create project from template: ${response.statusText}`);
    return response.json();
  }

  async getVideoGenerationQueue(): Promise<VideoGenerationQueue> {
    const response = await fetch(`${this.baseUrl}/api/video-studio/queue`);
    if (!response.ok) throw new Error(`Failed to fetch video generation queue: ${response.statusText}`);
    return response.json();
  }
}

// Singleton instance
export const videoStudioService = new VideoStudioServiceImpl();

// React hook
export const useVideoStudioService = (): VideoStudioService => {
  return videoStudioService;
};

// Helper functions
export function createStoryboardScene(
  title: string,
  description: string,
  prompt: string,
  durationSeconds: number,
  options?: Partial<StoryboardScene>
): StoryboardScene {
  return {
    id: `scene_${Date.now()}`,
    title,
    description,
    durationSeconds,
    prompt,
    style: options?.style || 'photorealistic',
    referenceImages: options?.referenceImages || [],
    cameraAngle: options?.cameraAngle || 'medium_shot',
    transition: options?.transition || 'cut',
    metadata: options?.metadata || {},
  };
}

export function calculateVideoDuration(storyboard: StoryboardScene[]): number {
  return storyboard.reduce((total, scene) => total + scene.durationSeconds, 0);
}

export function estimateGenerationTime(durationSeconds: number, quality: 'low' | 'medium' | 'high' = 'medium'): {
  estimatedMinutes: number;
  costCredits: number;
} {
  const qualityMultipliers = {
    low: { time: 0.5, credits: 1 },
    medium: { time: 1, credits: 2 },
    high: { time: 2, credits: 5 },
  };

  const multiplier = qualityMultipliers[quality];
  const baseTimePerSecond = 0.5; // 30 seconds per minute of video

  return {
    estimatedMinutes: Math.ceil((durationSeconds * baseTimePerSecond * multiplier.time) / 60),
    costCredits: Math.ceil(durationSeconds * multiplier.credits),
  };
}

export function validateStoryboard(storyboard: StoryboardScene[]): {
  valid: boolean;
  errors: string[];
  warnings: string[];
} {
  const errors: string[] = [];
  const warnings: string[] = [];

  if (storyboard.length === 0) {
    errors.push('Storyboard cannot be empty');
    return { valid: false, errors, warnings };
  }

  storyboard.forEach((scene, index) => {
    if (!scene.title.trim()) {
      errors.push(`Scene ${index + 1}: Title is required`);
    }

    if (!scene.description.trim()) {
      warnings.push(`Scene ${index + 1}: Description is empty`);
    }

    if (!scene.prompt.trim()) {
      errors.push(`Scene ${index + 1}: Prompt is required`);
    }

    if (scene.durationSeconds <= 0) {
      errors.push(`Scene ${index + 1}: Duration must be greater than 0`);
    }

    if (scene.durationSeconds > 60) {
      warnings.push(`Scene ${index + 1}: Duration exceeds 60 seconds, consider splitting`);
    }

    if (scene.prompt.length > 1000) {
      warnings.push(`Scene ${index + 1}: Prompt is very long, may affect generation quality`);
    }
  });

  const totalDuration = calculateVideoDuration(storyboard);
  if (totalDuration > 300) { // 5 minutes
    warnings.push('Total video duration exceeds 5 minutes, generation may take longer');
  }

  return {
    valid: errors.length === 0,
    errors,
    warnings,
  };
}

export function optimizeStoryboardForGeneration(storyboard: StoryboardScene[]): StoryboardScene[] {
  return storyboard.map(scene => ({
    ...scene,
    prompt: enhancePrompt(scene.prompt, scene.style),
  }));
}

function enhancePrompt(originalPrompt: string, style?: string): string {
  let enhanced = originalPrompt;

  // Add quality and detail enhancements
  if (!enhanced.includes('high quality')) {
    enhanced += ', high quality, detailed';
  }

  if (!enhanced.includes('cinematic')) {
    enhanced += ', cinematic lighting';
  }

  // Add style-specific enhancements
  if (style) {
    switch (style) {
      case 'photorealistic':
        enhanced += ', photorealistic, 8k resolution';
        break;
      case 'anime':
        enhanced += ', anime style, vibrant colors';
        break;
      case 'documentary':
        enhanced += ', documentary style, natural lighting';
        break;
    }
  }

  return enhanced;
}

// Progress tracking helper
export function pollVideoGenerationStatus(
  projectId: string,
  onUpdate: (status: VideoGenerationStatus) => void,
  intervalMs: number = 3000,
  maxAttempts: number = 120
): Promise<VideoGenerationStatus> {
  let attempts = 0;

  return new Promise((resolve, reject) => {
    const poll = async () => {
      try {
        const status = await videoStudioService.getVideoGenerationStatus(projectId);
        onUpdate(status);

        if (status.status === 'completed' || status.status === 'failed') {
          resolve(status);
          return;
        }

        attempts++;
        if (attempts >= maxAttempts) {
          reject(new Error('Maximum polling attempts reached'));
          return;
        }

        setTimeout(poll, intervalMs);
      } catch (error) {
        reject(error);
      }
    };

    poll();
  });
}

// File validation
export function validateImageFile(file: File): { valid: boolean; error?: string } {
  const validTypes = ['image/jpeg', 'image/png', 'image/webp'];
  const maxSize = 10 * 1024 * 1024; // 10MB

  if (!validTypes.includes(file.type)) {
    return { valid: false, error: 'Only JPEG, PNG, and WebP images are supported' };
  }

  if (file.size > maxSize) {
    return { valid: false, error: 'Image size must be less than 10MB' };
  }

  return { valid: true };
}

File: jeantrail_project/src-tauri/src/video_studio.rs
// AI Video Studio / Colab Integration Module
use axum::{Json, extract::{Path, Query, State}, extract::Multipart};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use uuid::Uuid;
use chrono::{DateTime, Utc};
use std::collections::HashMap;

#[derive(Debug, Serialize, Deserialize, sqlx::FromRow)]
pub struct VideoProject {
    pub id: Uuid,
    pub user_id: Uuid,
    pub title: String,
    pub description: Option<String>,
    pub storyboard: Value,
    pub status: VideoProjectStatus,
    pub job_id: Option<String>,
    pub job_config: Value,
    pub result_urls: Vec<String>,
    pub progress: i32,
    pub error_message: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "varchar")]
pub enum VideoProjectStatus {
    #[serde(rename = "draft")]
    Draft,
    #[serde(rename = "processing")]
    Processing,
    #[serde(rename = "completed")]
    Completed,
    #[serde(rename = "failed")]
    Failed,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct StoryboardScene {
    pub id: String,
    pub title: String,
    pub description: String,
    pub duration_seconds: u32,
    pub prompt: String,
    pub style: Option<String>,
    pub reference_images: Vec<String>,
    pub camera_angle: Option<String>,
    pub transition: Option<String>,
    pub metadata: Value,
}

#[derive(Debug, Deserialize)]
pub struct CreateVideoProjectRequest {
    pub title: String,
    pub description: Option<String>,
    pub storyboard: Vec<StoryboardScene>,
    pub job_config: Value,
}

#[derive(Debug, Deserialize)]
pub struct UpdateVideoProjectRequest {
    pub title: Option<String>,
    pub description: Option<String>,
    pub storyboard: Option<Vec<StoryboardScene>>,
    pub status: Option<VideoProjectStatus>,
}

#[derive(Debug, Deserialize)]
pub struct GenerateVideoRequest {
    pub project_id: Uuid,
    pub config: Option<Value>,
}

pub async fn list_video_projects(
    Query(params): Query<ListVideoProjectsQuery>,
) -> Result<Json<Vec<VideoProject>>, axum::http::StatusCode> {
    let mock_projects = vec![
        VideoProject {
            id: Uuid::new_v4(),
            user_id: Uuid::new_v4(),
            title: "Sunset Beach Video".to_string(),
            description: Some("A beautiful sunset at the beach with waves".to_string()),
            storyboard: create_mock_storyboard(),
            status: VideoProjectStatus::Completed,
            job_id: Some("colab_job_123".to_string()),
            job_config: serde_json::json!({
                "model": "sdxl",
                "frames_per_second": 30,
                "duration_seconds": 120,
                "resolution": "1920x1080"
            }),
            result_urls: vec!["https://storage.example.com/videos/sunset_beach.mp4".to_string()],
            progress: 100,
            error_message: None,
            created_at: Utc::now(),
            updated_at: Utc::now(),
        }
    ];

    Ok(Json(mock_projects))
}

#[derive(Debug, Deserialize)]
pub struct ListVideoProjectsQuery {
    pub user_id: Option<Uuid>,
    pub status: Option<VideoProjectStatus>,
    pub limit: Option<i64>,
    pub offset: Option<i64>,
}

fn create_mock_storyboard() -> Value {
    serde_json::json!([
        {
            "id": "scene_1",
            "title": "Opening Scene",
            "description": "Wide shot of beach at sunset",
            "duration_seconds": 30,
            "prompt": "Beautiful sunset over calm ocean waves, sandy beach, golden hour lighting, cinematic",
            "style": "photorealistic",
            "reference_images": [],
            "camera_angle": "wide_shot",
            "transition": "fade_in",
            "metadata": {}
        },
        {
            "id": "scene_2",
            "title": "Wave Details",
            "description": "Close-up of waves crashing",
            "duration_seconds": 20,
            "prompt": "Close-up shot of ocean waves gently crashing on shore, detailed water droplets, serene",
            "style": "photorealistic",
            "reference_images": [],
            "camera_angle": "close_up",
            "transition": "cut",
            "metadata": {}
        }
    ])
}

pub async fn get_video_project(
    Path(id): Path<Uuid>,
) -> Result<Json<VideoProject>, axum::http::StatusCode> {
    // Fetch specific video project
    Err(axum::http::StatusCode::NOT_FOUND)
}

pub async fn create_video_project(
    Json(request): Json<CreateVideoProjectRequest>,
) -> Result<Json<VideoProject>, axum::http::StatusCode> {
    let storyboard_json = serde_json::to_value(&request.storyboard).unwrap_or_else(|_| serde_json::json!([]));
    
    let project = VideoProject {
        id: Uuid::new_v4(),
        user_id: Uuid::new_v4(), // Get from auth context
        title: request.title,
        description: request.description,
        storyboard: storyboard_json,
        status: VideoProjectStatus::Draft,
        job_id: None,
        job_config: request.job_config,
        result_urls: vec![],
        progress: 0,
        error_message: None,
        created_at: Utc::now(),
        updated_at: Utc::now(),
    };

    // Save to database
    Ok(Json(project))
}

pub async fn update_video_project(
    Path(id): Path<Uuid>,
    Json(request): Json<UpdateVideoProjectRequest>,
) -> Result<Json<VideoProject>, axum::http::StatusCode> {
    // Update video project in database
    Err(axum::http::StatusCode::NOT_FOUND)
}

pub async fn delete_video_project(
    Path(id): Path<Uuid>,
) -> Result<(), axum::http::StatusCode> {
    // Delete video project from database
    Ok(())
}

pub async fn generate_video(
    Json(request): Json<GenerateVideoRequest>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Get project details
    let project = get_video_project_by_id(request.project_id).await?;
    
    // Update project status to processing
    update_project_status(request.project_id, VideoProjectStatus::Processing).await?;
    
    // Submit job to Colab/Gradio
    let job_request = serde_json::json!({
        "notebook_url": std::env::var("COLAB_VIDEO_NOTEBOOK_URL").unwrap_or_else(|_| "https://colab.research.google.com/drive/default".to_string()),
        "parameters": {
            "storyboard": project.storyboard,
            "config": request.config.unwrap_or_else(|| serde_json::json!({}))
        },
        "timeout_minutes": 60
    });
    
    // Submit to integrations module
    let job_response = submit_colab_video_job(job_request).await?;
    
    // Update project with job ID
    update_project_job_id(request.project_id, &job_response["job_id"].as_str().unwrap_or("")).await?;
    
    Ok(Json(serde_json::json!({
        "success": true,
        "project_id": request.project_id,
        "job_id": job_response["job_id"],
        "estimated_duration_minutes": 45
    })))
}

async fn get_video_project_by_id(id: Uuid) -> Result<VideoProject, axum::http::StatusCode> {
    // Fetch from database
    Err(axum::http::StatusCode::NOT_FOUND)
}

async fn update_project_status(id: Uuid, status: VideoProjectStatus) -> Result<(), Box<dyn std::error::Error>> {
    // Update in database
    Ok(())
}

async fn update_project_job_id(id: Uuid, job_id: &str) -> Result<(), Box<dyn std::error::Error>> {
    // Update in database
    Ok(())
}

async fn submit_colab_video_job(job_request: Value) -> Result<Value, Box<dyn std::error::Error>> {
    // Call integrations module
    Ok(serde_json::json!({
        "job_id": format!("colab_video_{}", Uuid::new_v4()),
        "status": "submitted"
    }))
}

pub async fn get_video_generation_status(
    Path(project_id): Path<Uuid>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Get project status and job progress
    let status = serde_json::json!({
        "project_id": project_id,
        "status": "processing",
        "progress": 65,
        "current_step": "Generating scene 2 of 3",
        "estimated_remaining_minutes": 15,
        "job_id": "colab_job_123",
        "started_at": "2024-01-15T10:00:00Z",
        "updated_at": "2024-01-15T10:20:00Z",
        "error": null
    });
    
    Ok(Json(status))
}

pub async fn cancel_video_generation(
    Path(project_id): Path<Uuid>,
) -> Result<Json<Value>, axum::http::StatusCode> {
    // Cancel the video generation job
    update_project_status(project_id, VideoProjectStatus::Draft).await?;
    
    Ok(Json(serde_json::json!({
        "success": true,
        "message": "Video generation cancelled",
        "project_id": project_id
    })))
}

pub async fn upload_reference_image(
    Path(project_id): Path<Uuid>,
    mut multipart: Multipart,
) -> Result<Json<Value>, axum::http::StatusCode> {
    use axum::body::Bytes;
    
    let mut uploaded_files = Vec::new();
    
    while let Some(field) = multipart.next_field().await.map_err(|_| axum::http::StatusCode::BAD_REQUEST)? {
        let name = field.name().unwrap_or("file");
        
        if name.starts_with("reference_image_") {
            let data = field.bytes().await.map_err(|_| axum::http::StatusCode::BAD_REQUEST)?;
            let filename = field.file_name().unwrap_or("image.jpg");
            
            // Save file to storage
            let file_url = save_uploaded_image(&data, filename).await?;
            uploaded_files.push(file_url);
        }
    }
    
    Ok(Json(serde_json::json!({
        "success": true,
        "uploaded_files": uploaded_files,
        "project_id": project_id
    })))
}

async fn save_uploaded_image(data: &Bytes, filename: &str) -> Result<String, axum::http::StatusCode> {
    // Save to file system or cloud storage
    // Return URL
    let file_id = Uuid::new_v4();
    Ok(format!("https://storage.example.com/reference_images/{}_{}", file_id, filename))
}

pub async fn get_video_templates() -> Result<Json<Value>, axum::http::StatusCode> {
    let templates = serde_json::json!([
        {
            "id": "template_1",
            "name": "Product Showcase",
            "description": "Template for showcasing products with dynamic scenes",
            "scenes": [
                {
                    "title": "Product Intro",
                    "prompt": "Product introduction with smooth camera movement",
                    "duration_seconds": 15
                },
                {
                    "title": "Features",
                    "prompt": "Product features demonstration",
                    "duration_seconds": 30
                },
                {
                    "title": "Call to Action",
                    "prompt": "Final scene with call to action",
                    "duration_seconds": 10
                }
            ],
            "total_duration": 55,
            "style": "modern_commercial"
        },
        {
            "id": "template_2",
            "name": "Social Media Short",
            "description": "Template for short social media videos (60 seconds)",
            "scenes": [
                {
                    "title": "Hook",
                    "prompt": "Eye-catching opening scene",
                    "duration_seconds": 5
                },
                {
                    "title": "Main Content",
                    "prompt": "Main message or story",
                    "duration_seconds": 45
                },
                {
                    "title": "Outro",
                    "prompt": "Closing with branding",
                    "duration_seconds": 10
                }
            ],
            "total_duration": 60,
            "style": "social_media"
        }
    ]);
    
    Ok(Json(templates))
}

pub async fn create_project_from_template(
    Path(template_id): Path<String>,
    Json(request): Json<Value>,
) -> Result<Json<VideoProject>, axum::http::StatusCode> {
    // Get template and create project based on it
    let template = get_video_template_by_id(&template_id).await?;
    
    let project = VideoProject {
        id: Uuid::new_v4(),
        user_id: Uuid::new_v4(),
        title: request["title"].as_str().unwrap_or("New Video Project").to_string(),
        description: request["description"].as_str().map(|s| s.to_string()),
        storyboard: template["scenes"].clone(),
        status: VideoProjectStatus::Draft,
        job_id: None,
        job_config: serde_json::json!({
            "style": template["style"],
            "total_duration": template["total_duration"]
        }),
        result_urls: vec![],
        progress: 0,
        error_message: None,
        created_at: Utc::now(),
        updated_at: Utc::now(),
    };
    
    // Save to database
    Ok(Json(project))
}

async fn get_video_template_by_id(template_id: &str) -> Result<Value, axum::http::StatusCode> {
    // Get template from database or predefined list
    let templates = vec![
        serde_json::json!({
            "id": "template_1",
            "name": "Product Showcase",
            "scenes": [],
            "total_duration": 55,
            "style": "modern_commercial"
        })
    ];
    
    templates
        .into_iter()
        .find(|t| t["id"] == template_id)
        .ok_or(axum::http::StatusCode::NOT_FOUND)
}

pub async fn get_video_generation_queue() -> Result<Json<Value>, axum::http::StatusCode> {
    let queue = serde_json::json!({
        "active_jobs": 3,
        "queued_jobs": 7,
        "completed_today": 12,
        "average_processing_time_minutes": 45,
        "estimated_wait_time_minutes": 15,
        "active_jobs": [
            {
                "project_id": Uuid::new_v4(),
                "title": "Beach Sunset",
                "progress": 75,
                "started_at": "2024-01-15T10:00:00Z",
                "estimated_completion": "2024-01-15T10:45:00Z"
            },
            {
                "project_id": Uuid::new_v4(),
                "title": "Product Demo",
                "progress": 30,
                "started_at": "2024-01-15T10:15:00Z",
                "estimated_completion": "2024-01-15T11:00:00Z"
            }
        ]
    });
    
    Ok(Json(queue))
}